### ISSUE TITLE
Class notation

### ISSUE THREAD

**Issue by PierBover** [ryansolid/mobx-jsx#7]
Hey @ryansolid have you considered adding a class based component?

The problem with using functions is that you can't use decorators such as `@observable`. The [decorators proposal](https://github.com/tc39/proposal-decorators) is already in stage 2 so we are going to see a lot more of those in the future.

So instead of doing this:
```js
const App = () => {
  const state = observable({ counter: 0 });
  return <div>{( state.counter )}</div>;
};
```
With decorators and [class fields](https://github.com/tc39/proposal-class-fields) (which are stage 3) it would something like:
```js
class App extends Component {
  @observable counter = 0;

  render () {
    return <div>{( this.counter )}</div>;
  }
};
```

**Comment by ryansolid** [comment]
It seems possible. I read a really interesting article from Dan Abramov on the subject [How Does React Tell a Class from a Function](https://overreacted.io/how-does-react-tell-a-class-from-a-function/). Right now I do some shortcuts that don't make this as succinct (ie.. not relying on a createComponent method in all cases) but I think with a little work I could make it configurable and make the dom-expressions compiler output code that could support it adding the extra code overhead for libraries that wish to use this syntax. This seems a natural option for MobX though as decorators are popular here, so personal opinion aside, I imagine most MobX users would be willing to pay the cost.

I'm in the process of doing a big refactor of dom-expressions. When I'm done I will add opt in support for this and turn it on by default for mobx-jsx.

**Comment by PierBover** [comment]
Awesome, thanks a lot for your work.

**Comment by ryansolid** [comment]
Ok, latest version of the plugin really opened up what I can do with it. I now support MobX the way a MobX user would expect including class components and better handling of arrays.

Example: https://codesandbox.io/s/mobx-counterclasses-uz7g9

Released in v0.6.0
